### PR TITLE
Improved README and docs on how to use the lib (CRAFT-707).

### PR DESCRIPTION
**Pull request by facundobatista**
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: docs/tutorials.rst [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+          -t, --trace:  Show all information needed to trace internal
+                        behaviour
+    
+    For a summary of all commands, run 'example-app help --all'.
+    
+Time to run the command on a file, you should see the successful message::
+
+    12:00:03|facundo@camdeonita:~/canonical/craft-cli (improve-readme *)$ touch testfile
+    12:02:29|facundo@camdeonita:~/canonical/craft-cli (improve-readme *)$ ls testfile2
+    ls: no se puede acceder a 'testfile2': No existe el archivo o el directorio
```
Comment (cmatsuoka) [inline docs/tutorials.rst:158]: It seems that you didn't run this command with `LC_ALL=C` or equivalent.

File: docs/tutorials.rst [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    12:02:37|facundo@camdeonita:~/canonical/craft-cli (improve-readme *)$ ls testfile
+    testfile
+    12:02:39|facundo@camdeonita:~/canonical/craft-cli (improve-readme *)$ env/bin/python example-app.py unlink testfile 
+    File removed succesfully.                                                                                                                                                                            
+    12:02:52|facundo@camdeonita:~/canonical/craft-cli (improve-readme *)$ LANG=C ls testfile
```
Comment (cmatsuoka) [inline docs/tutorials.rst:163]: And in this command the locale setting leaked to the tutorial :)

File: README.md [hunk 3]
```diff
@@ -1,12 +1,45 @@
 # Craft CLI
 
-Command Line Client builder that follows Canonical's guidelines for a command line
-interface defined in https://discourse.ubuntu.com/c/design/cli-guidelines/62.
+A Command Line Client builder that follows Canonical's guidelines for a command line
+interface defined [in this forum post](https://discourse.ubuntu.com/c/design/cli-guidelines/62).
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply README.md:4]: Improved.

File: README.md [hunk 4]
```diff
@@ -1,12 +1,45 @@
...
+The library provides two main functionalities: 
+
+- a framework to define and execute application commands, which involves argument parsing and the provision of help texts
+
+- infrastructure to handle the terminal and present all the outputs for the different application needs
+
+
+# Documentation
+
+The documentation is in Read The Docs, [check it out](https://craft-cli.readthedocs.io).
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply README.md:15]: better, thanks

File: craft_cli/dispatcher.py [hunk 5]
```diff
@@ -23,11 +23,27 @@
...
 from craft_cli.helptexts import HelpBuilder
 
-# a helper to group some commands together
 CommandGroup = namedtuple("CommandGroup", "name commands")
+"""Definition of a command group.
+
+A list of these is what is passed to the ``Dispatcher`` to run commands as part
+of the application.
+
+:param name: identifier of the command group (to be used in helping texts).
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_cli/dispatcher.py:32]: yes, thanks

File: craft_cli/dispatcher.py [hunk 6]
```diff
@@ -23,11 +23,27 @@
...
 
-# global options: the name used internally, its type, short and long parameters, and help text
 GlobalArgument = namedtuple("GlobalArgument", "name type short_option long_option help_message")
+"""Definition of a global argument to be handled by the Dispatcher.
+
+:param name: identifier of the argument (the reference in the dictionary returned
+    by ``Dispatcher.pre_parse_args`` method)
+:param type: the argument type: ``flag`` for arguments that are set to ``True`` if
+    specified (``False`` by default), or ``option`` if a value is needed after it.
+:param short_option: the short form of the argument (a dash with a letter, e.g. ``-s``).
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_cli/dispatcher.py:43]: Nop. We could add this functionality, but so far all options must have short and long forms.

File: craft_cli/dispatcher.py [hunk 7]
```diff
@@ -23,11 +23,27 @@
...
+:param long_option: the long form of the argument (two dashes and a name, e.g. ``--secure``).
+:param help_message: the one-line text that describes the argument, for building the helping texts.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_cli/dispatcher.py:45]: yes

File: docs/explanations.rst [hunk 8]
```diff
@@ -0,0 +1,188 @@
+************
+Explanations
+************
+
+About the appropiate mode to initiate `emit`
+============================================
+
+The first mandatory parameter of the ``emit`` object is ``mode``, which controls the verboseness level for the system to start with.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:8]: I like it, but we need to reflect that the value you pass there is only the verboseness of the system when the system starts, but maybe the user will change it through `--verbose` or other options.

File: docs/explanations.rst [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+
+How Craft CLI manage the application logs
+=========================================
+
+Unless overriden when ``emit`` is initiated (see :ref:`how to do that <howto_other_logfile>`), the application logs will be managed by the Craft CLI library, according to the following rules:
+
+- one log file is always produced for each application run (only exposed to the user if the application ends in error or a verbose run was requested by ``--verbose`` or ``--trace``), naming the files with a timestamp so they are unique.
+
+- log files are located in a directory with the application name under the user's log directory
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:27]: Removed the period in the first one, thanks!

File: docs/explanations.rst [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    <app> <command> <command-parameter> --verbose
+
+The Dispatcher automatically provides the following global arguments, but more can be specified through the `extra_global_args` option (see :ref:`how to do that <howto_global_args>`):
+
+- ``-q`` / ``--quiet``: sets the ``emit`` output level to QUIET
+- ``-v`` / ``--verbose``: sets the ``emit`` output level to VERBOSE
+- ``-t`` / ``--trace``: sets the ``emit`` output level to TRACE
+- ``-h`` / ``--help``: provides a help text for the application or command
+
+On the other hand, each command can specify its own arguments parsing rules using the ``fill_parser`` method, which receives an `ArgumentParser <https://docs.python.org/dev/library/argparse.html>`_ with all its features for parsing a command line argument. The parsing result will be passed to the command on execution, as the ``parsed_args`` parameter of the ``run`` method.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:52]: Much better, thanks!

File: docs/explanations.rst [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    emit.init(EmitterMode.NORMAL, "example-app", "Starting example app v1.")
+
+
+After bootstrapping the library as shown before, and importing ``emit`` wherever is needed, all its usage is just sending information to the user. The following sections describe the different ways of doing that.
+
+
+Regular messages
+~~~~~~~~~~~~~~~~
+
+The ``message`` metod is for the final output of the running command.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:89]: Yes, thanks

File: docs/explanations.rst [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+E.g.::
+
+    emit.message("The meaning of life is 42.")
+
+
+Progress messages
+~~~~~~~~~~~~~~~~~
+
+The ``progress`` method is for all the progress messages intended to provide information that the machinery is running and doing what.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:105]: Rewrote it to:

```
The ``progress`` method is to present all the messages that provide information on what the application is currently doing.
```

File: docs/explanations.rst [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    def open_stream(self, text: str) -> _StreamContextManager:
+
+E.g.::
+
+    with emit.open_stream("Running ls") as stream:
+        subprocess.run(["ls", "-l"], stdout=stream, stderr=stream)
+
+
+How to easily test different combinations
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:173]: No, it's not about testing (the current title is misleading!), but about *trying* combinations.

I changed the title to `How to easily try different combinations`. What do you think?

File: docs/howtos.rst [hunk 14]
```diff
@@ -0,0 +1,139 @@
+*******
+HOW TOs
+*******
+
+.. _howto_other_logfile:
+
+Use a different logfile structure than the default
+==================================================
+
+To override the default management of application log files (`detailed here <FIXME>`_), a file path can be specified when initiating the `emit` object, using the ``log_filepath`` parameter::
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/howtos.rst:10]: Yes! And actually filled the FIXME, which was wrongly left over.

File: docs/tutorials.rst (hunk 1 above)
Comment (facundobatista) [reply docs/tutorials.rst:158]: good catch!! I actually forgot to "process" the whole block, updated

File: docs/tutorials.rst (hunk 2 above)
Comment (facundobatista) [reply docs/tutorials.rst:163]: yes, the whole block needs processing, updated.

File: docs/explanations.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply docs/explanations.rst:8]: Maybe "... controls the initial verboseness level of the system"?

File: docs/explanations.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply docs/explanations.rst]: I see in the body text that "combinations" refer to combinations of different message types, what do you think about "How to easily try different message types"? It seems that just leaving "combinations" in the title does not make it clear what kind of combinations we're talking about.

File: docs/explanations.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst:8]: Nice! yes, thanks.

File: docs/explanations.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (facundobatista) [reply docs/explanations.rst]: Used the title you suggested, and improved the first paragraph to explain more what it is about...

**Review by cmatsuoka — APPROVED** [review]
Thanks!

**Pull request merged**